### ISSUE TITLE
FunctionalBatchMetric doesn't pass batch size to BatchMetricCallback

### ISSUE THREAD

**Issue by MartSlaaf** [catalyst-team/catalyst#1390]
## 🐛 Bug Report
`FunctionalBatchMetric` wrapped with `BatchMetricCallback` fails to pass actual inputs to functions. Because `FunctionalBatchMetric.update_key_value` at [_functional_metric.py#L84](https://github.com/catalyst-team/catalyst/blob/678dc06eda1848242df010b7f34adb572def2598/catalyst/metrics/_functional_metric.py#L84) expects `batc_size` as the first argument, but `BatchMetricCallback. _update_value_metric` at [metric.py#L166](https://github.com/catalyst-team/catalyst/blob/678dc06eda1848242df010b7f34adb572def2598/catalyst/callbacks/metric.py#L166) fails to provide it.

### How To Reproduce
1. Define custom metric function computed from predictions and targets
2. Wrap it with `FunctionalBatchMetric`
3. Wrap the result with `BatchMetricCallback`
4. Pass it to the runner as callback.

 

#### Code sample
```python
from torch.utils.data import TensorDataset, DataLoader
from torch import nn, optim, FloatTensor
import numpy as np
from sklearn.model_selection import train_test_split
from catalyst.runners.runner import SupervisedRunner
from catalyst.metrics._functional_metric import FunctionalBatchMetric
from catalyst.callbacks.metric import BatchMetricCallback

# Defining metric

def custom_metric_function(prediction, label):
    return ((prediction - label)**2).mean()
metric = FunctionalBatchMetric(metric_fn=custom_metric_function, metric_key='custom')
callback = BatchMetricCallback(metric=metric, input_key='logits', target_key='targets')

# Generating dataset

X = np.random.randn(1_000, 10)
y = X.sum(1)

train_X, test_X, train_y, test_y = train_test_split(X, y)

train_set = TensorDataset(FloatTensor(train_X), FloatTensor(train_y))
test_set = TensorDataset(FloatTensor(test_X), FloatTensor(test_y))

train_loader = DataLoader(train_set, batch_size=32)
test_loader = DataLoader(test_set, batch_size=32)

# Initializing model, criterion and optimizer

model = nn.Linear(10, 1)
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters())

# Getting error

runner = SupervisedRunner()
runner.train(model=model,
             loaders={'train': train_loader, 'valid': test_loader},
             criterion=criterion, 
             optimizer=optimizer, 
             callbacks=[callback],
             verbose=True)
```

### Expected behavior
I expect it to work.
I guess, either in callback or in metric batch size should be derived from the input lengths?

### Environment
Please copy and paste the output from our environment collection script
```
Catalyst version: 21.12
PyTorch version: 1.10.0+cu113
Is debug build: No
CUDA used to build PyTorch: 11.3
TensorFlow version: N/A
TensorBoard version: 2.7.0

OS: Ubuntu 20.04.3 LTS
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
CMake version: version 3.16.3

Python version: 3.9
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000
GPU 2: NVIDIA RTX A5000
GPU 3: NVIDIA RTX A5000

Nvidia driver version: 495.29.05
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] catalyst==21.12
[pip3] efficientnet-pytorch==0.6.3
[pip3] numpy==1.20.3
[pip3] numpydoc==1.1.0
[pip3] segmentation-losses-pytorch==0.0.1
[pip3] segmentation-models-pytorch==0.2.1
[pip3] tensorboard==2.7.0
[pip3] tensorboard-data-server==0.6.1
[pip3] tensorboard-plugin-wit==1.8.0
[pip3] tensorboardX==2.2
[pip3] torch==1.10.0+cu113
[pip3] torchaudio==0.10.0+cu113
[pip3] torchvision==0.11.1+cu113
[conda] blas                      1.0                         mkl
[conda] catalyst                  21.12                    pypi_0    pypi
[conda] efficientnet-pytorch      0.6.3                    pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.20.3           py39hf144106_0
[conda] numpy-base                1.20.3           py39h74d4b33_0
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1
[conda] segmentation-losses-pytorch 0.0.1                    pypi_0    pypi
[conda] segmentation-models-pytorch 0.2.1                    pypi_0    pypi
[conda] tensorboard               2.7.0                    pypi_0    pypi
[conda] tensorboard-data-server   0.6.1                    pypi_0    pypi
[conda] tensorboard-plugin-wit    1.8.0                    pypi_0    pypi
[conda] tensorboardx              2.2                      pypi_0    pypi
[conda] torch                     1.10.0+cu113             pypi_0    pypi
[conda] torchaudio                0.10.0+cu113             pypi_0    pypi
[conda] torchvision               0.11.1+cu113             pypi_0    pypi
```

### Checklist
- [x] bug description
- [x] steps to reproduce
- [x] expected behavior
- [x] environment
- [x] code sample / screenshots
  
### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [ ] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

**Comment by MartSlaaf** [comment]
Hm, after some more time in sources, I have found the `FunctionalBatchMetricCallback`, which looks working good. Just couldn't find a description for it in the docs. 
From docs now it looks like somewhat logical, that callback shouldn't bother about is the metric functional or not. Only about batch/loader difference.
